### ISSUE TITLE
Fails to watch directory with exclamation mark in name (windows)

### ISSUE THREAD

**Issue by jdoose** [remy/nodemon#690]
nodemon fails to watch the files in a directory, if the directory name contains an exclamation mark. 
This happens under Windows, cannot confirm for other systems.
When the directory of the node js file contains a quotiation mark it reads " [nodemon] watching 0 files". If I rename the directory to the same name _without_ exclamation mark it works.

Using pure chokidar it works in both cases.

I guess it has something to do with parsing ignore/include patterns, but that is just a rough guess.

Nodemon version 1.7.1
It worked when using nodemon 1.3.x

Dump output

```
C:\doose-Data\Work\Raup_Ritter\CHECK!Lehrerzimmer\sub1>nodemon --dump --verbose --debug app.js
  nodemon bus new listener: reset (0) +0ms
  nodemon bus new listener: reset (0) +5ms
  nodemon bus new listener: quit (0) +16ms
  nodemon bus new listener: quit (0) +1ms
  nodemon bus new listener: restart (0) +2ms
  nodemon bus new listener: restart (0) +2ms
  nodemon bus new listener: reset (2) +12ms
  nodemon bus emit: reset +2ms
  nodemon:watch resetting watchers +1ms
  nodemon reset +2ms
  nodemon config: dirs +236ms [ 'C:\\doose-Data\\Work\\Raup_Ritter\\CHECK!Lehrerzimmer\\sub1' ]
18 Oct 11:42:59 - [nodemon] 1.7.1
18 Oct 11:42:59 - [nodemon] to restart at any time, enter `rs`
18 Oct 11:42:59 - [nodemon] ignoring: .git C:\doose-Data\Work\Raup_Ritter\CHECK!Lehrerzimmer\sub1\node_modules/**/* bower_components .sass-cache
18 Oct 11:42:59 - [nodemon] watching: *.*
18 Oct 11:42:59 - [nodemon] watching extensions: js
--------------
node: v0.12.6
nodemon: 1.7.1
command: node C:\Users\doose\AppData\Roaming\npm\node_modules\nodemon\bin\nodemon.js --dump --verbose --debug app.js
cwd: C:\doose-Data\Work\Raup_Ritter\CHECK!Lehrerzimmer\sub1
OS: win32 x64
--------------
{ run: false,
  system: { cwd: 'C:\\doose-Data\\Work\\Raup_Ritter\\CHECK!Lehrerzimmer\\sub1' },
  required: false,
  dirs: [ 'C:\\doose-Data\\Work\\Raup_Ritter\\CHECK!Lehrerzimmer\\sub1' ],
  timeout: 1000,
  options:
   { dump: true,
     verbose: true,
     ignore:
      [ '.git',
        'node_modules',
        'bower_components',
        '.sass-cache',
        re: /\.git|node_modules|bower_components|\.sass\-cache/ ],
     watch: [ '*.*', re: /.*\..*/ ],
     ignoreRoot: [ '.git', 'node_modules', 'bower_components', '.sass-cache' ],
     restartable: 'rs',
     colours: true,
     execMap: { py: 'python', rb: 'ruby' },
     stdin: true,
     runOnChangeOnly: false,
     stdout: true,
     execOptions:
      { script: 'app.js',
        exec: 'node',
        args: [ '--debug' ],
        scriptPosition: 1,
        nodeArgs: undefined,
        ext: 'js',
        env: {},
        execArgs: [] },
     monitor:
      [ '*.*',
        '!.git',
        '!C:\\doose-Data\\Work\\Raup_Ritter\\CHECK!Lehrerzimmer\\sub1\\node_modules/**/*',
        '!bower_components',
        '!.sass-cache' ] },
  load: [Function],
  reset: [Function: reset],
  lastStarted: 0,
  loaded: [],
  watchInterval: null,
  command:
   { raw: { executable: 'node', args: [ '--debug', 'app.js' ] },
     string: 'node --debug app.js' } }
--------------
18 Oct 11:43:00 - [nodemon] exiting
```

**Comment by remy** [comment]
Gosh yes...

I don't have a windows machine to debug this, do you think you can try to
isolate this with the chokidar watching module?

**Comment by jdoose** [comment]
Hi! I already took a look into it but was not able to figure anything out.
But sure, I will have another look at it and trying to debug it.

I set up a quick app that uses chokidar to recurse through all sub dirs, basically it was the sample from the chokidar github page. That was working, so I guess it has something to do with the options nodemon is passing.

I'm afraid I will not be able to look into this in the next few days, but I think it should be possible next weekend!

**Comment by barczaG** [comment]
+1 on OS X

**Comment by grokky1** [comment]
Still a problem. I'm on Win10 x64 with the latest node and npm and nodemon.

My path has a `!` so nodemon shows `"watching 0 files"`. 😢 

Any workaround other than to rename the path?

**Comment by jarpi** [comment]
Hi, 
I would like to work on this fix, actually I've [raised a PR](https://github.com/remy/nodemon/pull/1137) with the failing test cases as stated in the contribution guidelines.
As part of the investigation, I've found that some test cases are not properly working, ie: the test should fail but it isn't. Should I open a new issue or can we handle as a dependency of this one?

**Comment by remy** [comment]
Probably a new issue. I'm thinking of (somehow) putting this on appveyor (I
think) so it can run Windows tests too. But I'd need help, and I'm pretty
sure the failing tests you're seeing is related to Windows OS.

**Comment by remy** [comment]
Can you (all) try with current latest nodemon? It disables chokidar's globbing by default on Windows, so hopefully fixes this issue in one go (…hopefully).

**Comment by jarpi** [comment]
Hello, 
So far I've tested with Ubuntu and I can say +1.
Just raised a PR with the failing test case, but it seems it's unable to build the project. (configuration?)
The issues with failing tests, are related to false positives (I think)
I'm just working on a clean branch [to get CI under Windows with appveyor]
(https://github.com/jarpi/nodemon/tree/windows_compliant) :), the tests need to be adapted to Windows specifities.

**Comment by mannypamintuan** [comment]
I tested this on macOSx and issue still remains.

I used the following structure to manually test:
https://github.com/mannypamintuan/testNodemon

And I am able to watch files where either the file does not contain `!` or the directory path does not contain `!`.

```
--------------
node: v8.9.1
nodemon: 1.12.5
command: /Users/../.nvm/versions/node/v8.9.1/bin/node /Users/../.nvm/versions/node/v8.9.1/bin/nodemon --dump --verbose --debug test!directory/testFile.js
cwd: /Users/../dev/test/testNodemon
OS: darwin x64
--------------
{ run: false,
  system: { cwd: '/Users/../dev/test/testNodemon' },
  required: false,
  dirs: [ '/Users/../dev/test/testNodemon' ],
  timeout: 1000,
  options:
   { dump: true,
     verbose: true,
     ignore:
      [ '.git',
        '.nyc_output',
        '.sass-cache',
        'bower_components',
        'coverage',
        'node_modules',
        re: /\.git|\.nyc_output|\.sass\-cache|bower_components|coverage|node_modules/ ],
     watch: [ '*.*', re: /.*\..*/ ],
     ignoreRoot:
      [ '.git',
        '.nyc_output',
        '.sass-cache',
        'bower_components',
        'coverage',
        'node_modules' ],
     restartable: 'rs',
     colours: true,
     execMap: { py: 'python', rb: 'ruby' },
     stdin: true,
     runOnChangeOnly: false,
     stdout: true,
     watchOptions: {},
     execOptions:
      { script: 'test!directory/testFile.js',
        exec: 'node',
        args: [ '--debug' ],
        scriptPosition: 1,
        nodeArgs: undefined,
        execArgs: [],
        ext: 'js,json',
        env: {} },
     monitor:
      [ '*.*',
        '!/Users/../dev/test/testNodemon/.git/**/*',
        '!.nyc_output',
        '!.sass-cache',
        '!bower_components',
        '!coverage',
        '!node_modules' ] },
  signal: 'SIGUSR2',
  load: [Function],
  reset: [Function: reset],
  lastStarted: 0,
  loaded: [ '/Users/../dev/test/testNodemon/package.json' ],
  watchInterval: null,
  command:
   { raw:
      { executable: 'node',
        args: [ '--debug', 'test!directory/testFile.js' ] },
     string: 'node --debug test!directory/testFile.js' } }
--------------
[nodemon] exiting
```

**Comment by dominykas** [comment]
This bug is caused by `is-glob@2.x` (used by `chokidar > anymatch > micromatch`) - it treats all `!` characters as glob characters, whereas only the leading `!` should be considered a glob. It seems to be fixed in `is-glob@4.x` - https://github.com/micromatch/is-glob/blob/master/index.js#L25 (although I'm still not sure how it would deal with folders that actually start with a `!`). 

`micromatch@3.x > is-glob@4.x` is out, but `anymatch` still uses `minimatch@2.x` - a quick hack locally seemed to make folder `test!test` watchable locally (macOS).

**Comment by dominykas** [comment]
anymatch already has a PR open: https://github.com/es128/anymatch/pull/23

**Comment by remy** [comment]
@jdoose I believe your particular problem is fixed on Windows as nodemon now by default disables globbing (which is why I'm closing the original issue).

Equally, I can't replicate this at all on Windows… I've used both git bash and the command prompt with and older version of nodemon:

![nodemon-bang](https://user-images.githubusercontent.com/13700/34262468-60b324fc-e664-11e7-854a-b8eaf7074a80.png)

Others - there's an active workaround, but again, I can't replicate on a Mac either (note that I've tested this against 1.12.x _and_ 1.7.x and it still works) - @mannypamintuan - I was using our repo as the example, so let me know if I've run it wrong or something (there wasn't a readme, I just went by the command from your info dump above).

<img width="603" alt="screen shot 2017-12-21 at 15 34 41" src="https://user-images.githubusercontent.com/13700/34262525-88746280-e664-11e7-9156-087527a82d9e.png">

I'm also tracking the es128/anymatch#23 PR and once that's in and updated in Chokidar I'll push an update to the `package-lock.json`.
